refactor: Update schema readme. (#108)
y-scope/log-surgeon: docs/schema.md

Author: Reflowed

// Custom variables
hex:[a-fA-F]+
hasNumber:.*\d.*

> equalsCapture:.*=(?<equals>.*[a-zA-Z0-9].*)

Reviewer: Can a variable regex contain multiple capture groups? Can it contain nested capture groups?

Author: It can contain multiple capture groups. I'm not 100% sure what happens with nested capture groups, in theory it should be fine, but I haven't tested it.

```
* `delimiters: \t\r\n:,!;%'` indicates that ` `, `\t`, `\r`, `\n`, `:`, `,`,
  `!`, `;`, `%`, and `'` are delimiters.
* `timestamp` matches two different patterns:
    * 2023-04-19 12:32:08.064
    * [20230419-12:32:08]
* `int`, `float`, `hex`, `hasNumber`, and `equalsCapture` all match different
  user defined variables.
* `equalsCapture` also contains the named capture group `equals`. This allows
  the user to extract the substring following the equals sign.

> ## Regular Expression Syntax

Author: Added at bottom.

// Custom variables
hex:[a-fA-F]+
hasNumber:.*\d.*
equalsCapture:.*=(?<equals>.*[a-zA-Z0-9].*)
```

> * `delimiters: \t\r\n:,!;%'` indicates that ` `, `\t`, `\r`, `\n`, `:`, `,`,
>   `!`, `;`, `%`, and `'` are delimiters.

Reviewer 2: The actual example doesn't have a single quote character `'`.
```suggestion
* `delimiters: \t\r\n:,!;%` indicates that ` `, `\t`, `\r`, `\n`, `:`, `,`, `!`, `;`, and `%` are 
  delimiters.
```

(I reflowed the text as well.)

* `timestamp` matches two different patterns:
  * `2023-04-19 12:32:08.064`
  * `[20230419-12:32:08]`
* `int`, `float`, `hex`, `hasNumber`, and `equalsCapture` all match different user defined
  variables.
* `equalsCapture` also contains the named capture group `equals`. This allows the user to extract
  the substring following the equals sign.

## Regular Expression Syntax

The following regular expression rules are supported by the schema. When building a regular
expression, the rules are applied as they appear in this list, from top to bottom.

```txt
REGEX RULE   DEFINITION
ab           Match 'a' followed by 'b'

> a|b          Match a OR b

Reviewer 2: ```suggestion
a|b          Match 'a' OR 'b'
```

a+           Match 'a' 1 or more times
a{N}         Match 'a' exactly N times
a{N,M}       Match 'a' between N and M times

> \d           Match any digit 0-9
> \s           Match any whitespace character (' ', '\r', '\t', '\v', or '\f')
> .            Match any *non-delimiter* character
> (abc)        Subexpression (concatenates abc)
> (?<name>abc) Named capture group (matches 'abc' and saves it as 'name')
> 
> * Special characters include: ( ) * + , - . [ \ ] ^ { | } < > ?
>   - These must be escaped when used literally in patterns.

Reviewer 2: I'm not really sure what the best way to make this consistent is.

Every existing entry beings with 'Match' so we should either stick with that or drop it. New entries have a "name" and then the description in parentheses, while the old definitions have no "name" (e.g. the subexpression parens say "concatenates", but we never actually named concatenation earlier).
(I can't make a suggestion to add names like concatenation/alternation/Kleene star because github counts it outside the diff.)

Also, the new special character list is space separated for visual clarity, but the whitespace characters use quotations.
```suggestion
\d           Match any digit: 0-9
\s           Match any whitespace character: ' ' \r \t \v \f
.            Match any *non-delimiter* character
(ab)         Match a subexpression (matches 'ab')
(?<name>ab)  Match and save a named capture group (matches 'ab' and saves it as 'name')

- Special characters the requiring escaping: ( ) * + , - . [ \ ] ^ { | } < > ?
  - e.g. 'a\*' will literally match 'a*'
  - Note: the list is space separated, but space itself is not a special character.
```

Maybe use the suggestion for some inspiration. I also changed abc -> ab to match the first definition. 
Side questions:
* Just double checking, we don't supported unnamed groups right?
* On a side note, why does `,` need to be escaped?
